feat(#1949): Preparing for CHT 5.0, remove existing upgrades dir (#1953)
medic/cht-docs: content/en/hosting/cht/migration/preparing-for-5.md

* Have [backups]({{< relref "/hosting/cht/docker/backups/" >}}) that have been tested to work
* Ensure your [CHT Conf]({{< relref "/community/contributing/code/cht-conf/" >}}) version is up to date
* Review the change log to understand how end users might be affected

## CouchDB Nouveau

{{< callout type="info" >}} Applies to: [Kubernetes]({{< relref "/hosting/cht/kubernetes/" >}}) hosted CHT instances {{< /callout >}}

A major feature of CHT 5.0 is replacing the old freetext search in 4.x with CouchDB Nouveau. The addition of Nouveau [decreases the amount of disk space](https://github.com/medic/cht-core/issues/9898#issuecomment-2864545914) used by the CHT.  Disk space has been shown to be a major contributor to hosting costs, hence Nouveau is the first feature to [reduce Hosting Total Cost of Ownership (TCO)](https://github.com/medic/cht-roadmap/issues/171).

Kubernetes based deployments will need to follow the migration steps from the deprecated [CHT 4.x Helm Charts repository](https://github.com/medic/helm-charts/) to the new [CHT 5.x Helm Charts](https://github.com/medic/cht-core/tree/master/scripts/build/helm) in the main CHT repository.

> Please see the technical guide on how to migrate Kubernetes based deployments to 5.0.

Author: yup yup! I was using `tk` [placeholders](https://en.wikipedia.org/wiki/To_come_(publishing)), but agree that `todo` is more a code native thing (and markdown/hugo is code)

fixed in  dd553bbe

* The Nouveau index data on the server will be stored in `${COUCHDB_DATA}/nouveau` for single-node CouchDBs and in `${DB1_DATA}/nouveau` for clustered CouchDBs.
* The following `medic-client` views no longer exist. Be sure to update any custom scripts which use them:  `contacts_by_freetext`,  `contacts_by_type_freetext` and  `reports_by_freetext` .

## Upgrade service removed from Kubernetes

{{< callout type="info" >}} Applies to: [Kubernetes]({{< relref "/hosting/cht/kubernetes/" >}}) hosted CHT instances {{< /callout >}}

In CHT 4.x there are two ways to upgrade a Kubernetes hosted CHT instance:

1. Click "Stage" and then "Upgrade" in the CHT administration interface.  
2. Push a new helm chart to Kubernetes that specified different versions.

A recurring problem seen on production deployments was that, over time, CHT administrators would use the CHT administration interface to upgrade.  Then, during a migration or a restore from backup, Kubernetes administrators would inadvertently downgrade a CHT instance by using an out of date Helm chart.

> To avoid this situation, the upgrade service has been removed from Kubernetes based deployments.

Author: good question!  @dianabarsan do you know what we'll do? Are we going to hide #1 or hide buttons like #2?

<img width="1231" height="799" alt="hidden upgrade" src="https://github.com/user-attachments/assets/1bb29a2f-2b27-44ce-aa27-a8f36877bb1a" />

> To avoid this situation, the upgrade service has been removed from Kubernetes based deployments.

Reviewer: The page will just show the "stage" button. 

<img width="1389" height="412" alt="Image" src="https://github.com/user-attachments/assets/9aef8589-8a16-415b-b456-a90ed5b2a21c" />

Background information:
* [Hide upgrade button in admin app for k8s deployments, while still allowing staging upgrades](https://github.com/medic/cht-core/issues/9954)
* [Remove upgrade service from Helm charts](https://github.com/medic/cht-core/issues/10186)

## Token login requires `app_url`

{{< callout type="info" >}} Applies to: Instances that use [token login]({{< relref "/building/login/#magic-links-for-logging-in-token-login" >}}) {{< /callout >}}

For instances that use token login,  be sure declare `app_url` in your [settings file]({{< relref "/building/reference/app-settings/token_login/#app_settingsjson-token_login" >}}). This is backwards compatible and safe to do while still on CHT 4.x.  

Be sure to use [the new declarative config](#declarative-configuration) as well.

Background information:
* [Require `app_url` to be set when enabling `token_login`](https://github.com/medic/cht-core/issues/9983)

> ## Enabling languages via generated docs

Reviewer: > @dianabarsan what is the failure behavior here for these cases?

When the `languages` property is not set, all languages will become available. we have a migration for existing server configurations, but someone can overwrite them.

### Declarative Configuration 

{{< callout type="info" >}} Applies to: All Instances {{< /callout >}}

<!-- 
    todo: add missing TK content on how to make a declarative config
-->
tk - all the content ;)  - but upgrade your app settings to be declarative.

Background information:
* [Make declarative config mandatory](https://github.com/medic/cht-core/issues/5906)

### Increase ecmaVersion ES linting to version 6

{{< callout type="info" >}} Applies to: CHT versions between 4.0 and 4.4  {{< /callout >}}

Deployments running CHT Version 4.0 and 4.4 must upgrade to version 4.5 or later before upgrading to version 5.x.

Background information:
* [Increase ecmaversion linting for ddocs](https://github.com/medic/cht-core/issues/9202)

### Ensure booleans are used with signoff fields

{{< callout type="info" >}} Applies to: Instances where `needs_signoff` triggers replication of reports  {{< /callout >}}

In some scenarios where `needs_signoff` is set to string `"false"` instead of boolean `false`, replication will happen when it shouldn't.  This is because the string of `"false"` is a truthy value.  

> CHT 5.0 now correctly enforces this to be a boolean.  Deployments to make sure that workflows that use `needs_signoff` field have a boolean instead of a string.

Reviewer 2: hi @mrjones-plip thanks for documenting this
so technically the field **in the xlsx form** can be a string or a boolean eg needs_signoff -> true or needs_signoff -> "true" basically a boolean or "boolean string" 
the report generated will have a string result either way
not sure if it matters much here though but just wanted to point out that it should work either way

Author: Thanks for the tip @jonathanbataire !  I think the text is still accurate as is, but feel free to make a suggestion if you think it needs improvement.  I very much welcome more eyes on this doc! <3

---

> title: "Preparing to upgrade to CHT 5.0"

Reviewer 3: ```suggestion
title: "Preparing to Upgrade to CHT 5.0"
```

linkTitle: "Preparing for 5.0"

> weight: 

Reviewer 3: suggestion: Let's move this guide before the "Preparing for 4.0" in the section hierarchy.

Author: good thinking! fixed in d6b12065 

turns out the helm chart migration doc didn't have frontmatter?  I added it tweaked the  title a bit

Background information:
* [Add documentation for migrating from 4.x medic/helm-charts to the new production charts in 5.x cht-core](https://github.com/medic/cht-docs/issues/1943)
* [Hide upgrade button in admin app for k8s deployments, while still allowing staging upgrades](https://github.com/medic/cht-core/issues/9954)
* [Remove upgrade service from Helm charts](https://github.com/medic/cht-core/issues/10186)

### Token login requires `app_url`

{{< callout type="info" >}} Applies to: Instances that use [token login]({{< relref "/building/login/#magic-links-for-logging-in-token-login" >}}) {{< /callout >}}

For instances that use token login,  be sure declare `app_url` in your [settings file]({{< relref "/building/reference/app-settings/token_login/#app_settingsjson-token_login" >}}). This is backwards compatible and safe to do while still on CHT 4.x.

Background information:
* [Require `app_url` to be set when enabling `token_login`](https://github.com/medic/cht-core/issues/9983)

### Enabling languages via generated docs

{{< callout type="info" >}} Applies to: Instances that enable languages via web interface {{< /callout >}}

> As of [CHT 4.2.0]({{< relref "/releases/4_2_0/#technical-improvements" >}}), deployment administrators can enable or disable supported languages [via the settings file]({{< relref "/building/reference/app-settings/#app_settingsjson" >}}). CHT 5.0 removes this web interface to enable or disable languages and requires this to be done via App Settings.

Author: I think just removing this last sentence makes sense then?  If not , please unresolve and suggest some better wording - I struggled a bit with this!

removed in 6dcdf613

CHT 5.0 now correctly enforces this to be a boolean.  Deployments to make sure that workflows that use `needs_signoff` field have a boolean instead of a string.

Background information:
* [Replication fails to filter out reports with `needs_signoff` set to false](https://github.com/medic/cht-core/issues/10183)

### Angular 20 requires Chrome 107 or later

{{< callout type="info" >}} Applies to: all instances  {{< /callout >}}

CHT 5.0 is preparing fo an upgraded version of [Angular](https://blog.angular.dev/announcing-angular-v20-b5c9c06cf301). As this version of Angular will require Chrome 107 or later, we're making this a requirement in 5.0 as a breaking change to pave the path for the upgrade later. When [researching the impact of this upgrade](https://github.com/medic/cht-core/issues/10029#issuecomment-3358338361) we surveyed over 100,000 end users devices and found that less than 0.5% of users are affected.  These users are running Chrome 106 released 3 years ago at this writing. A best practice is to have users always run the latest version if possible.

> To check if any of your users are impacted, check the [user-devices API](/building/reference/api/#get-apiv2exportuser-devices). This is an authenticated API endpoint and will return JSON for all users for all time.  As this may include multiple entries per user, it's important to filter out duplicate users and of course filter out users who are running Chrome 107 or later. 

Reviewer 3: ```suggestion
To check if any of your users are impacted, check the [user-devices API](/building/reference/api/#get-apiv2exportuser-devices). This is an authenticated API endpoint and will return JSON for all users for all time.  As this may include multiple entries per user, it's important to filter out duplicate users and those running Chrome 107 or later. 
```

```shell
jq -r '.[] |  select(.apk|length> 0) 
  | select(.browser.name == "Chrome")
  | select(.date > "2025-04-01")
  | "\(.deviceId) \(.user) \(.date) \(.browser.version)"' \
  devices.json | sort -u -k1,2r \
  | sort -u -k1,1 | cut -f2,3,4 -d' '| sort -rh -k3 \
  | sed 's/ /,/g' > devices.csv
```

The resulting file `devices.csv` will have all of your users sorted by the version of Chrome they use.  Scroll to the bottom and find any users running a version lower than `107` and ensure they update to a newer version.  The easiest way to upgrade is to open the Play Store and update any out of date software, being sure to update [WebView](https://play.google.com/store/apps/details?id=com.google.android.webview&hl=en).

> For instances with users who share devices, or have multiple devices they log in on, they may have multiple entries. Thi is because the CHT creates one entry per device per user. Be sure to upgrade all affected devices!

Reviewer 3: ```suggestion
For instances with users who share devices, or have multiple devices they log in on, they may have multiple entries. This is because the CHT creates one entry per device per user. Be sure to upgrade all affected devices!
```

Background information:
* [Upgrade to Angular 20](https://github.com/medic/cht-core/issues/10029)

## Non-Breaking Changes

> These changes apply to all deployments.  No specific action needs to be taken, but administrators should be aware of the changes.

Reviewer 3: ```suggestion
{{< callout >}}
These changes apply to all deployments.  No specific action is required, but administrators should be aware of the changes.
{{< /callout >}}
```

### CouchDB Nouveau

> A major feature of CHT 5.0 is replacing the old freetext search in 4.x with CouchDB Nouveau. The addition of Nouveau [decreases the amount of disk space](https://github.com/medic/cht-core/issues/9898#issuecomment-2864545914) used by the CHT.  Disk space has been shown to be a major contributor to hosting costs, hence Nouveau is the first feature to [reduce Hosting Total Cost of Ownership (TCO)](https://github.com/medic/cht-roadmap/issues/171).

Reviewer 3: ```suggestion
A major feature of CHT 5.0 is replacing the freetext search in 4.x with CouchDB Nouveau. The addition of Nouveau [decreases the amount of disk space](https://github.com/medic/cht-core/issues/9898#issuecomment-2864545914) used by the CHT.  Disk space is a major contributor to hosting costs, hence Nouveau is a feature meant to [reduce Hosting Total Cost of Ownership (TCO)](https://github.com/medic/cht-roadmap/issues/171).
```

Nouveau has the following impact on all CHT deployments upgrading to 5.0:

> * Offline users' devices will have their search indexes rebuilt after the upgrade. During the rebuild, search functionality will be initially unavailable. However, once the indexing is complete, the search behavior will be unchanged from before for all users:  the search experience is the same in both CHT 4.x and CHT 5.x.

Reviewer 3: ```suggestion
* Offline users' devices will have their search indexes rebuilt after the upgrade. During the rebuild, search functionality will be initially unavailable. However, once indexing is complete, the search behavior will remain unchanged for all users: the search experience is the same in both CHT 4.x and CHT 5.x.
```

---
title: "Preparing to upgrade to CHT 5.0"
linkTitle: "Preparing for CHT 5.0"
weight: 
description: >
  Steps to ensure your CHT App will run smoothly on CHT 5.0 and later
  
---

## Introduction

CHT releases use [Semantic Versioning](https://en.wikipedia.org/wiki/Semver#Semantic_versioning) (also known as "SemVer"), which means that the CHT upgrade from the major 4.x version to the 5.x version denotes there are breaking changes. The key to a successful upgrade will be to understand and plan for these breaking changes.

While this releases breaking changes mainly apply to Kubernetes based deployments (e.g. Nouveau & Upgrade service), deployments using token login may be affected as well.

As with all upgrades, major or minor, you should always:

> * Have [backups](/hosting/cht/docker/backups/) that have been tested to work

Reviewer 4: Do we still need to use the `ref` shortcode for local links?  

```
[backups]({{% ref "hosting/cht/docker/backups" %}})
```

Either way, there are a number of `localhost` links below that I think need fixed too.

Author: huh - yeah - you're totally right about using ref and relref!  I thought I remember seeing that hugo updates and hextra didn't require it, but checking our style guide and they're still there! Thanks for noticing!

Fixed use of links and removing localhost in 28c8a806

To avoid this situation, the upgrade service has been removed from Kubernetes based deployments.

Kubernetes based deployments will need to follow the migration steps from the deprecated [CHT 4.x Helm Charts repository](https://github.com/medic/helm-charts/) to the new [CHT 5.x Helm Charts](https://github.com/medic/cht-core/tree/master/scripts/build/helm) in the main CHT repository. Please see the TK technical guide on how to migrate Kubernetes based deployments to 5.0.

Background information:
* [Hide upgrade button in admin app for k8s deployments, while still allowing staging upgrades](https://github.com/medic/cht-core/issues/9954)
* [Remove upgrade service from Helm charts](https://github.com/medic/cht-core/issues/10186)

## Token login requires `app_url`

{{< callout type="info" >}} Applies to: Instances that use [token login](/building/login/#magic-links-for-logging-in-token-login) {{< /callout >}}

> For instances that use token login,  be sure declare `app_url` in your [settings file](/building/reference/app-settings/token_login/#app_settingsjson-token_login). This is backwards compatible and safe to do while still on CHT 4.x.  

Reviewer 4: Good call on the "backwards compatible" note! 👍

Background information:
* [Require `app_url` to be set when enabling `token_login`](https://github.com/medic/cht-core/issues/9983)

## Enabling languages via generated docs

{{< callout type="info" >}} Applies to: Instances that enable languages via web interface {{< /callout >}}

> As of CHT 4.20, deployment administrators can enable or disable supported languages [via the settings file](/building/reference/app-settings/#app_settingsjson). CHT 5.0 removes this web interface to enable or disable languages and requires this to be done via App Settings.

Reviewer 4: > As of CHT 4.20

Is this note about `4.20` correct?  I don't see anything in the [release notes](https://docs.communityhealthtoolkit.org/releases/4_20_0) about languages.

Author: D'oh!  I miss-read that release - thanks for double checking.  It's `4.2.0` and it IS [mentioned there](https://docs.communityhealthtoolkit.org/releases/4_2_0/) :sweat_smile: 

fixed in 91e494c2
